clients/src/main/java/org/apache/kafka/clients/consumer/internals/AssignorSelection.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Selection of a type of assignor used by a member to get partitions assigned as part of a
+ * consumer group. Selection could be one of:
+ * <li>CLIENT assignors</li>
+ * <li>SERVER assignors</li>
+ * <p/>
+ * Client assignors include of a list of
+ * {@link org.apache.kafka.common.message.ConsumerGroupHeartbeatRequestData.Assignor}
+ * Server assignors include a name of the server assignor selected, ex. uniform, range.
+ */
+public class AssignorSelection {
+    public enum Type { CLIENT, SERVER }
```

Comment:
I think the type declaration is redundant because we already know which assignor is used by checking their existence.

Also if we use polymorphic constructor, we don't need to pass null's around.

```
   private Optional<String> serverAssignor = Optional.empty();
    private Optional<List<ConsumerGroupHeartbeatRequestData.Assignor>> clientAssignors = Optional.empty();

    private AssignorSelection(List<ConsumerGroupHeartbeatRequestData.Assignor> assignors) {
        this.clientAssignors = Optional.of(assignors);
    }

    private AssignorSelection(String serverAssignor) {
        this.serverAssignor = Optional.of(serverAssignor);
    }
```

Reply:
Regarding the type, I added it explicitly thinking of cases where we might be interested in checking the type without actually retrieving the assignors (ex. validations). For better readability, seems better to be able to do `if (selection.type() == AssignorSelection.Type.CLIENT)` 

Regarding the constructors, having static factory methods here instead of constructors allows to have meaningful names, so clients using the AssignorSelection can explicitly indicate what they want to create (`newClientAssignors` or `newServerAssignor`). We only pass null internally to the private constructor so seemed like a fair price to pay internally in the class, in order to provide a very explicit way to create new instances.